public/app/plugins/datasource/loki/LanguageProvider.ts

```diff
@@ -461,4 +472,18 @@ export default class LokiLanguageProvider extends LanguageProvider {
 
     return labelValues ?? [];
   }
+
+  async getLogInfo(selector: string): Promise<{ extractedLabelKeys: string[]; hasJSON: boolean; hasLogfmt: boolean }> {
+    let series = [];
```

Comment:
@matyax nice work! (i know this has changed since this comment, i checked the actual version)

so, `extractedLabelKeys`... i assume the question is: from the data-samples, i'd like to extract all the label-keys, correct?

you will need to go through the dataframe-array returned by `datasource.getDataSamples`. in loki dataframes, there is a field (it's a "column" of data), named "labels", and it's contents are javascript-objects that contain the key-value pairs for the given row.

so something like:
```ts
const allKeysDuplicated = series.forEach(frame => {
   const labelsField = frame.fields.find(f => f.name === 'labels');
   const objs = labelsField.values.toArray();
   const keys = objs.map(obj => Object.keys(obj));
   return keys;
});
const allKeys = Array.from(new Set(allKeysDuplicated));
```

i wrote this blindly, i hope it explains what i mean, if not just ping me :)